src/kvstore/plugins/elasticsearch/ESListener.cpp

```diff
@@ -0,0 +1,272 @@
+                    << lastApplyLogFile_->c_str() << "\" (" << errno << "): "
+                    << strerror(errno);
+        close(fd);
+        return false;
+    }
+    close(fd);
+    return true;
+}
+
+std::string
+ESListener::encodeAppliedId(LogID lastId, TermID lastTerm, LogID lastApplyLogId) const noexcept {
+    std::string val;
+    val.reserve(sizeof(LogID) * 2 + sizeof(TermID));
+    val.append(reinterpret_cast<const char*>(&lastId), sizeof(LogID))
+       .append(reinterpret_cast<const char*>(&lastTerm), sizeof(TermID))
+       .append(reinterpret_cast<const char*>(&lastApplyLogId), sizeof(LogID));
+    return val;
+}
+
+bool ESListener::appendDocItem(std::vector<DocItem>& items, const KV& kv, int32_t vIdLen) const {
+    auto isEdge = NebulaKeyUtils::isEdge(vIdLen, kv.first);
+    return isEdge ? appendEdgeDocItem(items, kv, vIdLen) : appendTagDocItem(items, kv, vIdLen);
+}
+
+bool ESListener::appendEdgeDocItem(std::vector<DocItem>& items,
+                                   const KV& kv,
+                                   int32_t vIdLen) const {
+    auto edgeType = NebulaKeyUtils::getEdgeType(vIdLen, kv.first);
+    auto schema = schemaMan_->getEdgeSchema(spaceId_, edgeType);
+    if (schema == nullptr) {
+        VLOG(3) << "get edge schema failed, edgeType " << edgeType;
+        return false;
+    }
+    auto reader = RowReaderWrapper::getEdgePropReader(schemaMan_,
+                                                      spaceId_,
+                                                      edgeType,
+                                                      kv.second);
+    if (reader == nullptr) {
+        VLOG(3) << "get edge reader failed, schema ID " << edgeType;
+        return false;
+    }
+    return appendDocs(items, schema.get(), reader.get(), edgeType, true);
+}
+
+bool ESListener::appendTagDocItem(std::vector<DocItem>& items,
+                                  const KV& kv,
+                                  int32_t vIdLen) const {
+    auto tagId = NebulaKeyUtils::getTagId(vIdLen, kv.first);
+    auto schema = schemaMan_->getTagSchema(spaceId_, tagId);
+    if (schema == nullptr) {
+        VLOG(3) << "get tag schema failed, tagId " << tagId;
+        return false;
+    }
+    auto reader = RowReaderWrapper::getTagPropReader(schemaMan_,
+                                                     spaceId_,
+                                                     tagId,
+                                                     kv.second);
+    if (reader == nullptr) {
+        VLOG(3) << "get tag reader failed, tagID " << tagId;
+        return false;
+    }
+    return appendDocs(items, schema.get(), reader.get(), tagId, false);
+}
+
+bool ESListener::appendDocs(std::vector<DocItem>& items,
+                            const meta::SchemaProviderIf* schema,
+                            RowReader* reader,
+                            int32_t schemaId,
+                            bool isEdge) const {
+    auto spaceName = getSpaceName();
```

Comment:
We don't need to get space name for each DocItem?

Reply:
> We don't need to get space name for each DocItem?

In this clip. All DocItems have the same space name, so we don't need to get space name for each DocItem. WDYT?

Reply:
If so, we could add a interface in `Listener` to init.

Reply:
> If so, we could add a interface in `Listener` to init.

Aha, I also think it's a good idea.